### PR TITLE
Add CLI support for Multi-cluster engine secret:

### PR DESCRIPTION
**Pull request by jnpacker**
* Documentation
  * Secret support:
    * Base Domain
    * AWS Key ID
    * AWS Secret Key 
    * SSH Keys Public/private 
    * Pull-Secret

This allows existing MCE credentials or new credentials created via the MCE console to be consumed by the CLI. This is a known quantity in MCE that is documented and utilized by MCE/ACM users

Signed-off-by: Joshua Packer <jpacker@redhat.com>

**What this PR does / why we need it**:
  1. Allows the hypershift CLI to leverage MCE created Credential Secrets for the AWS provider.
  2. This provides an easy repeatable provisioning flow with less chance of a typo
  3. This helps the MCE Console generate/describe a hypershift cli command to create a cluster
  4. Users can create secret to represent a number of parameters the CLI uses, having the secret ties nicely into a tekton task for GitOps and other run scenarios.
 
**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #
https://issues.redhat.com/browse/HOSTEDCP-640

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [x] This change includes docs. 
- [ ] ~This change includes unit tests.~

### CODE REVIEW

File: cmd/cluster/cluster.go [hunk 1]
```diff
@@ -69,8 +69,6 @@ func NewCreateCommands() *cobra.Command {
 	cmd.PersistentFlags().BoolVar(&opts.Wait, "wait", opts.Wait, "If the create command should block until the cluster is up. Requires at least one node.")
 	cmd.PersistentFlags().DurationVar(&opts.Timeout, "timeout", opts.Timeout, "If the --wait flag is set, set the optional timeout to limit the waiting duration. The format is duration; e.g. 30s or 1h30m45s; 0 means no timeout; default = 0")
 
-	cmd.MarkPersistentFlagRequired("pull-secret")
```
Comment (jnpacker) [inline cmd/cluster/cluster.go:72]: Because I removed pull-secret at the cluster level, it needed to be inserted into each platform/create.go file (except AWS and eventually Azure).

File: cmd/cluster/aws/create.go [hunk 2]
```diff
@@ -55,6 +54,21 @@ func NewCreateCommand(opts *core.CreateOptions) *cobra.Command {
...
+		if util.IsEmpty(opts.CredentialSecretName) {
+			if err := util.IsRequiredOption("aws-creds", opts.AWSPlatform.AWSCredentialsFile); err != nil {
+				return err
+			}
+			if err := util.IsRequiredOption("pull-secret", opts.PullSecretFile); err != nil {
+				return err
+			}
+		} else {
+			//Check the secret exists now, otherwise stop
+			opts.Log.Info("Retreiving credentials secret", "namespace", opts.Namespace, "name", opts.CredentialSecretName)
```
Comment (jnpacker) [inline cmd/cluster/aws/create.go:66]: This check is done now instead of when we use the secret data, as it allows for a clear error message when the secret is missing. Otherwise the error is wrapped in a different error message and is much harder to read.

File: cmd/cluster/core/create.go [hunk 3]
```diff
@@ -164,14 +168,25 @@ func createCommonFixture(ctx context.Context, opts *CreateOptions) (*apifixtures
...
 	if len(opts.ControlPlaneOperatorImage) > 0 {
 		annotations[hyperv1.ControlPlaneOperatorImageAnnotation] = opts.ControlPlaneOperatorImage
 	}
 
-	pullSecret, err := ioutil.ReadFile(opts.PullSecretFile)
-	if err != nil {
-		return nil, fmt.Errorf("failed to read pull secret file: %w", err)
+	var pullSecret []byte
+	var err error
+	if util.IsEmpty(opts.PullSecretFile) && util.IsNotEmpty(opts.CredentialSecretName) {
```
Comment (jnpacker) [inline cmd/cluster/core/create.go:180]: This allows the --pull-secret parameter to override the secret, otherwise use the secret data.

**Review by rokej — APPROVED** [review]
Looks good to me.

**Comment by jnpacker** [comment]
Review and comment requested in #project-hypershift
https://coreos.slack.com/archives/C01C8502FMM/p1669302067388539

&
 
https://coreos.slack.com/archives/C01C8502FMM/p1669126226464099

File: cmd/cluster/aws/create.go [hunk 4]
```diff
@@ -44,8 +44,7 @@ func NewCreateCommand(opts *core.CreateOptions) *cobra.Command {
 	cmd.Flags().StringVar(&opts.AWSPlatform.EndpointAccess, "endpoint-access", opts.AWSPlatform.EndpointAccess, "Access for control plane endpoints (Public, PublicAndPrivate, Private)")
 	cmd.Flags().StringVar(&opts.AWSPlatform.EtcdKMSKeyARN, "kms-key-arn", opts.AWSPlatform.EtcdKMSKeyARN, "The ARN of the KMS key to use for Etcd encryption. If not supplied, etcd encryption will default to using a generated AESCBC key.")
 	cmd.Flags().BoolVar(&opts.AWSPlatform.EnableProxy, "enable-proxy", opts.AWSPlatform.EnableProxy, "If a proxy should be set up, rather than allowing direct internet access from the nodes")
-
-	cmd.MarkFlagRequired("aws-creds")
+	cmd.Flags().StringVar(&opts.CredentialSecretName, "secret-creds", opts.CredentialSecretName, "A kubernete's secret with a platform credential, pull-secret and base-domain. The secret must exist in the supplied \"--namespace\"")
```
Comment (derekwaynecarr) [inline cmd/cluster/aws/create.go:47]: typo: Kubernetes

File: docs/content/how-to/aws/create-aws-hosted-cluster-secret-credentials.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+apiVersion: v1
+kind: Secret
+type: Opaque
+metadata:
+  name: my-aws-credentials
+  namespace: clusters
+  labels:
+    cluster.open-cluster-management.io/credentials: ""
+    cluster.open-cluster-management.io/type: aws
+stringData:
```
Comment (derekwaynecarr) [inline docs/content/how-to/aws/create-aws-hosted-cluster-secret-credentials.md:23]: what is the historical context for keeping the base domain and the ssh info in this same secret?

**Review by derekwaynecarr — COMMENTED** [review]
seems useful to me.

only question was historical context on what goes in or out of the secret going forward.

File: cmd/cluster/agent/create.go [hunk 6]
```diff
@@ -29,6 +29,7 @@ func NewCreateCommand(opts *core.CreateOptions) *cobra.Command {
 	cmd.Flags().StringVar(&opts.AgentPlatform.APIServerAddress, "api-server-address", opts.AgentPlatform.APIServerAddress, "The API server address that should be used for components outside the control plane")
 	cmd.Flags().StringVar(&opts.AgentPlatform.AgentNamespace, "agent-namespace", opts.AgentPlatform.AgentNamespace, "The namespace in which to search for Agents")
 	cmd.MarkFlagRequired("agent-namespace")
+	cmd.MarkPersistentFlagRequired("pull-secret")
```
Comment (enxebre) [inline cmd/cluster/agent/create.go:32]: why is this needed and it wasn't before?

File: cmd/cluster/aws/create.go (hunk 4 above)
Comment (enxebre) [reply cmd/cluster/aws/create.go:47]: >credential, pull-secret and base-domain

In which contractual format? which keys in the cm?

File: cmd/cluster/core/create.go [hunk 7]
```diff
@@ -118,21 +120,22 @@ type KubevirtPlatformCreateOptions struct {
...
-	Region             string
-	RootVolumeIOPS     int64
-	RootVolumeSize     int64
-	RootVolumeType     string
-	EndpointAccess     string
-	Zones              []string
-	EtcdKMSKeyARN      string
-	EnableProxy        bool
+	AWSCredentialsFile      string
+	AWSCredentialSecretName string
```
Comment (enxebre) [inline cmd/cluster/core/create.go:124]: I can see where AWSCredentialSecretName is consumed?

File: cmd/cluster/core/create.go [hunk 8]
```diff
@@ -143,6 +146,8 @@ type AzurePlatformOptions struct {
 	AvailabilityZones []string
 }
 
+const managedClusterNameKey = "cluster.open-cluster-management.io/managedcluster-name"
```
Comment (enxebre) [inline cmd/cluster/core/create.go:149]: please move const declaration to the top of the file and document it, rename to managedClusterNameAnnotation and document it, e.g
```// managedClusterNameAnnotation annotates clusters....```

Also

File: cmd/cluster/core/create.go [hunk 9]
```diff
@@ -164,14 +169,25 @@ func createCommonFixture(ctx context.Context, opts *CreateOptions) (*apifixtures
 		k, v := pair[0], pair[1]
 		annotations[k] = v
 	}
+	// Add annotation to support import into MCE
```
Comment (enxebre) [inline cmd/cluster/core/create.go:172]: why is this annotation being added unconditionally?
We have an --annotation flag in the cli that enable set annotations freely

File: docs/content/how-to/aws/create-aws-hosted-cluster-secret-credentials.md (hunk 5 above)
Comment (jnpacker) [reply docs/content/how-to/aws/create-aws-hosted-cluster-secret-credentials.md:23]: ACM/MCE hive provisioning, this has been the existing flow for a while, both in the console and the automated flows. We have asks to support multiple secrets, but this is not from a majority of our user base.

We are looking at revisiting this at some point in the future:
1. In the console flows should we allow mulitple credentials to be referenced and each reference only pulls the relevant keys.
2. Generate NEW credentials in ACM using existing credentials as a base (clone and edit)
3. Double check to see if there have been more then the one or two RFE's to justify the first two.
4. In our anecdotal data, for automated flows this often simplifies the number of secrets to manage and setup once.

File: cmd/cluster/core/create.go (hunk 3 above)
Comment (enxebre) [reply cmd/cluster/core/create.go:181]: why do we want that?
can we just simplify and if CredentialSecretName use that and ignore everything else?

Comment (enxebre) [reply cmd/cluster/core/create.go:181]: If we still want the fallback I'd avoid the if else, e.g.
```
	if opts.CredentialSecretName != "" {
		pullSecret, err = util.GetDockerConfigJSON(opts.CredentialSecretName, opts.Namespace)
		if err != nil {
			return nil, err
		}
	} 
	
        // overrides secret if set
	if opts.PullSecretFile != "" {
		pullSecret, err = ioutil.ReadFile(opts.PullSecretFile)
		if err != nil {
			return nil, fmt.Errorf("failed to read pull secret file: %w", err)
		}
	}
```

File: cmd/cluster/core/create.go [hunk 10]
```diff
@@ -188,6 +204,20 @@ func createCommonFixture(ctx context.Context, opts *CreateOptions) (*apifixtures
 		if err != nil {
 			return nil, fmt.Errorf("failed to generate ssh keys: %w", err)
 		}
+	} else if util.IsNotEmpty(opts.CredentialSecretName) {
```
Comment (enxebre) [inline cmd/cluster/core/create.go:207]: Can we be consistent with above and do
```
	if opts.CredentialSecretName != "" {
		...
	}

	// overrides secret if set
	if opts.SSHKeyFile != "" {
		...
	}
```

File: cmd/cluster/kubevirt/create.go [hunk 11]
```diff
@@ -40,6 +40,8 @@ func NewCreateCommand(opts *core.CreateOptions) *cobra.Command {
 	cmd.Flags().StringVar(&opts.KubevirtPlatform.ContainerDiskImage, "containerdisk", opts.KubevirtPlatform.ContainerDiskImage, "A reference to docker image with the embedded disk to be used to create the machines")
 	cmd.Flags().StringVar(&opts.KubevirtPlatform.ServicePublishingStrategy, "service-publishing-strategy", opts.KubevirtPlatform.ServicePublishingStrategy, fmt.Sprintf("Define how to expose the cluster services. Supported options: %s (Use LoadBalancer and Route to expose services), %s (Select a random node to expose service access through)", IngressServicePublishingStrategy, NodePortServicePublishingStrategy))
 
+	cmd.MarkPersistentFlagRequired("pull-secret")
```
Comment (enxebre) [inline cmd/cluster/kubevirt/create.go:43]: why is this needed?

File: cmd/cluster/agent/create.go (hunk 6 above)
Comment (jnpacker) [reply cmd/cluster/agent/create.go:32]: When I removed `cmd.MarkPersistentFlagRequired("pull-secret") from core, so I re-inserted it in each of the refrences except AWS and eventually Azure when support gets added.

File: cmd/cluster/core/create.go (hunk 7 above)
Comment (jnpacker) [reply cmd/cluster/core/create.go:124]: Good catch. This was from an earlier iteration.

File: cmd/cluster/core/create.go (hunk 8 above)
Comment (jnpacker) [reply cmd/cluster/core/create.go:149]: done

File: cmd/cluster/core/create.go (hunk 9 above)
Comment (jnpacker) [reply cmd/cluster/core/create.go:172]: This helps with MCE import and general usability with MCE, but could just be documented as you suggest.

This annotation can be used to import the HostedCluster with a value other then that of the HostedCluster.Name.
For example, you could set this value to the infra-id when importing into MCE.

I'll remove it for now for simplicity.

File: cmd/cluster/core/create.go (hunk 3 above)
Comment (jnpacker) [reply cmd/cluster/core/create.go:181]: I would have preferred not to do the fallback and just always use the credential, but I got the feedback from the team and @derekwaynecarr that on the CLI folks might want to mix and match.

I would rather the simple flow with no overrides. Just do a check to see if they specified --base-domain, pull-secret or sshkey with --secret-creds and if that is the case, throw an error.

File: cmd/cluster/core/create.go (hunk 10 above)
Comment (jnpacker) [reply cmd/cluster/core/create.go:207]: done

File: cmd/cluster/core/create.go [hunk 12]
```diff
@@ -412,10 +442,21 @@ func defaultNetworkType(ctx context.Context, opts *CreateOptions, releaseProvide
 }
 
 func getReleaseSemanticVersion(ctx context.Context, opts *CreateOptions, provider releaseinfo.Provider, readFile func(string) ([]byte, error)) (*semver.Version, error) {
-	pullSecretBytes, err := readFile(opts.PullSecretFile)
-	if err != nil {
-		return nil, fmt.Errorf("cannot read pull secret file %s: %w", opts.PullSecretFile, err)
+	var pullSecretBytes []byte
+	var err error
+	if util.IsEmpty(opts.PullSecretFile) && util.IsNotEmpty(opts.CredentialSecretName) {
```
Comment (jnpacker) [inline cmd/cluster/core/create.go:447]: Adjusted the override like above.

File: cmd/cluster/kubevirt/create.go (hunk 11 above)
Comment (jnpacker) [reply cmd/cluster/kubevirt/create.go:43]: Same as mentioned above, it was removed from core so I pushed it down into the individual deploys.

File: cmd/util/util.go [hunk 13]
```diff
@@ -0,0 +1,26 @@
+package util
```
Comment (jnpacker) [inline cmd/util/util.go:1]: I removed this file.

Instead of IsEmpty and IsNotEmpty, I switched back to len(var) == 0 and len(var) > 0, this conforms with the rest of the CLI code.

File: cmd/util/util.go [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+// IsNotEmpty is easier to read then len(value) > 0)
+func IsNotEmpty(value string) bool {
+	return len(value) > 0
+}
+
+func IsEmpty(value string) bool {
+	return !IsNotEmpty(value)
+}
+
+func IsRequiredOption(flag string, value string) error {
```
Comment (jnpacker) [inline cmd/util/util.go:20]: I moved this method to `isRequiredOption` in the aws package where it can be called by aws/create.go and aws/destroy.go

File: cmd/cluster/aws/create.go (hunk 4 above)
Comment (jnpacker) [reply cmd/cluster/aws/create.go:47]: What about this:
> "A Kubernetes secret with a platform credential (--aws-creds), --pull-secret and --base-domain value. The secret must exist in the supplied \"--namespace\""

File: cmd/cluster/core/create.go (hunk 3 above)
Comment (jnpacker) [reply cmd/cluster/core/create.go]: Suggestion implemented and still supports overrides

File: cmd/cluster/aws/create.go [hunk 15]
```diff
@@ -55,6 +54,21 @@ func NewCreateCommand(opts *core.CreateOptions) *cobra.Command {
...
 
+		if len(opts.CredentialSecretName) == 0 {
+			if err := isRequiredOption("aws-creds", opts.AWSPlatform.AWSCredentialsFile); err != nil {
+				return err
+			}
+			if err := isRequiredOption("pull-secret", opts.PullSecretFile); err != nil {
+				return err
+			}
+		} else {
+			//Check the secret exists now, otherwise stop
```
Comment (enxebre) [inline cmd/cluster/aws/create.go:65]: nit: comments by convention leave space after and end with a period.
e.g `// Check the secret exists now, otherwise stop.`

**Comment by enxebre** [comment]
/approve
needs rebase.
PTAL cc @csrwng @sjenning

**Comment by jnpacker** [comment]
Rebased.
